externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td

```diff
@@ -14,22 +14,96 @@ include "torch-mlir-dialects/Dialect/Tcp/IR/TcpBase.td"
 
 include "mlir/IR/OpBase.td"
 
-def Tcp_DummyOp : Tcp_Op<"dummy", []> {
-  let summary = "dummy op";
+def Tcp_TanhOp : Tcp_UnaryElementwiseOp<"tanh"> {
+  let summary = "Computes tanh of input, elementwise";
 
   let description = [{
-    Dummy identity op that returns the input as output
+    Computes elementwise hyperbolic tangent of the input tensor.
   }];
 
   let arguments = (ins
-    Tcp_Tensor:$in
+    Tcp_FloatOrComplexTensor:$in
   );
 
   let results = (outs
-    Tcp_Tensor:$out
+    Tcp_FloatOrComplexTensor:$out
   );
 
   let assemblyFormat = "$in attr-dict `:` type($in) `->` type($out)";
 }
 
+def Tcp_AddOp : Tcp_BinaryElementwiseOp<"add", [SameOperandsAndResultElementType]> {
+  let summary = "Computes elementwise addition";
+
+  let description = [{
+    Computes the elementwise addition of `in1` and `in2`.
+
+    `in1` and `in2` must have the same shape and data type.
+
+    A tcp.add operation will pass verification `in1`, `in2` and the result have
+    compatible shapes.  Since TCP tensors have static ranks and element types,
+    this means:
+
+     * `in1`, `in2` and `out` must have the same rank.
+     * `in1`, `in2` and `out` must have the same element type.
+     * `in1`, `in2` and `out` must have the same size or a `?` on each dim.
+  }];
+
+  let arguments = (ins
+    Tcp_Tensor:$in1,
+    Tcp_Tensor:$in2
+  );
+
+  let results = (outs
+    Tcp_Tensor:$out
+  );
+
+  let assemblyFormat = "$in1 `,` $in2 attr-dict `:` type($in1) `,` type($in2) `->` type($out)";
+}
+
+def Tcp_BroadcastOp : Tcp_Op<"broadcast", [
+          AllRanksMatch<["in", "out"]>,
+          AllElementTypesMatch<["in", "out"]>,
+          PredOpTrait<
+            "attribute `axes` must be in increasing order",
+            CPred<"::llvm::is_sorted(getAxes(), "
```

Comment:
I'd move this into utility functions and just call those from here instead of this in tablegen.